### PR TITLE
Bug 1419650 - Use type inference where possible to remove unneeded static type specifications

### PR DESCRIPTION
**Pull request by adamnemecek**

### CODE REVIEW

File: Client/Application/AppDelegate.swift [hunk 1]
```diff
@@ -125,7 +125,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
 
         // Temporary fix for Bug 1390871 - NSInvalidArgumentException: -[WKContentView menuHelperFindInPage]: unrecognized selector
         if #available(iOS 11, *) {
-            if let clazz = NSClassFromString("WKCont" + "ent" + "View"), let swizzledMethod = class_getInstanceMethod(TabWebViewMenuHelper.self, #selector(TabWebViewMenuHelper.swizzledMenuHelperFindInPage)) {
+            if let clazz = NSClassFromString("WKCont" + "ent" + "View"), let swizzledMethod = class_getInstanceMethod(TabWebViewMenuHelper.self, #selector(swizzledMenuHelperFindInPage)) {
```
Comment (farhanpatel) [inline Client/Application/AppDelegate.swift]: This is causing the build to fail. I dont think using `swizzledMenuHelperFindInPage` here will work

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 2]
```diff
@@ -58,9 +58,9 @@ class BrowserViewController: UIViewController {
 
     lazy fileprivate var customSearchEngineButton: UIButton = {
         let searchButton = UIButton()
-        searchButton.setImage(UIImage(named: "AddSearch")?.withRenderingMode(.alwaysTemplate), for: UIControlState())
-        searchButton.addTarget(self, action: #selector(BrowserViewController.addCustomSearchEngineForFocusedElement), for: .touchUpInside)
-        searchButton.accessibilityIdentifier = "BrowserViewController.customSearchEngineButton"
```
Comment (farhanpatel) [inline Client/Frontend/Browser/BrowserViewController.swift]: The accessibilityIdentifier here should not change

**Comment by farhanpatel** [comment]
@adamnemecek Your PR is setup to run in the CI. You might not be able to see the log but the pass/fail should help.

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 2 above)
Comment (adamnemecek) [reply Client/Frontend/Browser/BrowserViewController.swift]: super sorry about that. should be good now.

File: Client/Frontend/AuthenticationManager/PasscodeViews.swift [hunk 3]
```diff
@@ -166,8 +166,8 @@ class PasscodePane: UIView {
             make.size.equalTo(PasscodeUX.PasscodeFieldSize)
         }
         layoutIfNeeded()
-        NotificationCenter.default.addObserver(self, selector: #selector(PasscodePane.keyboardWillHide(_:)), name: NSNotification.Name.UIKeyboardWillHide, object: nil)
-        NotificationCenter.default.addObserver(self, selector: #selector(PasscodePane.keyboardWillShow(_:)), name: NSNotification.Name.UIKeyboardWillShow, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillHide), name: .UIKeyboardWillHide, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow), name: .UIKeyboardWillShow, object: nil)
```
Comment (jhugman) [inline Client/Frontend/AuthenticationManager/PasscodeViews.swift]: Nit: unsure I like this or not.

File: Client/Frontend/Browser/ButtonToast.swift [hunk 4]
```diff
@@ -39,7 +39,7 @@ class ButtonToast: UIView {
     }()
     fileprivate var animationConstraint: Constraint?
     fileprivate lazy var gestureRecognizer: UITapGestureRecognizer = {
-        let gestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(ButtonToast.handleTap(_:)))
+        let gestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(ButtonToast.handleTap))
```
Comment (jhugman) [inline Client/Frontend/Browser/ButtonToast.swift]: Nit: why leave `ButtonToast` here?

File: Client/Frontend/Browser/TabToolbar.swift [hunk 5]
```diff
@@ -65,30 +65,30 @@ open class TabToolbarHelper: NSObject {
...
-        let longPressGestureStopReloadButton = UILongPressGestureRecognizer(target: self, action: #selector(TabToolbarHelper.didLongPressStopReload(_:)))
+        let longPressGestureStopReloadButton = UILongPressGestureRecognizer(target: self, action: #selector(TabToolbarHelper.didLongPressStopReload))
         toolbar.stopReloadButton.addGestureRecognizer(longPressGestureStopReloadButton)
-        toolbar.stopReloadButton.addTarget(self, action: #selector(TabToolbarHelper.didClickStopReload), for: UIControlEvents.touchUpInside)
+        toolbar.stopReloadButton.addTarget(self, action: #selector(TabToolbarHelper.didClickStopReload), for: .touchUpInside)
 
         toolbar.tabsButton.addTarget(self, action: #selector(TabToolbarHelper.didClickTabs), for: .touchUpInside)
-        let longPressGestureTabsButton = UILongPressGestureRecognizer(target: self, action: #selector(TabToolbarHelper.didLongPressTabs(_:)))
+        let longPressGestureTabsButton = UILongPressGestureRecognizer(target: self, action: #selector(TabToolbarHelper.didLongPressTabs))
```
Comment (jhugman) [inline Client/Frontend/Browser/TabToolbar.swift]: Nit: remove `TabToolbarHelper`?

File: Client/Frontend/Home/HomePanelViewController.swift [hunk 6]
```diff
@@ -120,7 +120,7 @@ class HomePanelViewController: UIViewController, UITextFieldDelegate, HomePanelD
         updateButtons()
 
         // Gesture recognizer to dismiss the keyboard in the URLBarView when the buttonContainerView is tapped
-        let dismissKeyboardGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(HomePanelViewController.dismissKeyboard(_:)))
+        let dismissKeyboardGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(HomePanelViewController.dismissKeyboard))
```
Comment (jhugman) [inline Client/Frontend/Home/HomePanelViewController.swift]: Nit: remove `HomePanelViewController`?

File: Client/Frontend/Home/ReaderPanel.swift [hunk 7]
```diff
@@ -191,8 +191,8 @@ class ReadingListPanel: UITableViewController, HomePanel {
 
     init() {
         super.init(nibName: nil, bundle: nil)
-        NotificationCenter.default.addObserver(self, selector: #selector(ReadingListPanel.notificationReceived(_:)), name: NotificationFirefoxAccountChanged, object: nil)
-        NotificationCenter.default.addObserver(self, selector: #selector(ReadingListPanel.notificationReceived(_:)), name: NotificationDynamicFontChanged, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(ReadingListPanel.notificationReceived), name: NotificationFirefoxAccountChanged, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(ReadingListPanel.notificationReceived), name: NotificationDynamicFontChanged, object: nil)
```
Comment (jhugman) [inline Client/Frontend/Home/ReaderPanel.swift]: Nit: remove `ReadingListPanel`?

File: Client/Frontend/Home/RecentlyClosedTabsPanel.swift [hunk 8]
```diff
@@ -84,7 +84,7 @@ class RecentlyClosedTabsPanelSiteTableViewController: SiteTableViewController {
     weak var recentlyClosedTabsPanel: RecentlyClosedTabsPanel?
 
     fileprivate lazy var longPressRecognizer: UILongPressGestureRecognizer = {
-        return UILongPressGestureRecognizer(target: self, action: #selector(RecentlyClosedTabsPanelSiteTableViewController.longPress(_:)))
+        return UILongPressGestureRecognizer(target: self, action: #selector(RecentlyClosedTabsPanelSiteTableViewController.longPress))
```
Comment (jhugman) [inline Client/Frontend/Home/RecentlyClosedTabsPanel.swift]: Nit: remove `RecentlyClosedTabsPanelSiteTableViewController`?

File: Client/Frontend/Home/RemoteTabsPanel.swift [hunk 9]
```diff
@@ -479,7 +479,7 @@ fileprivate class RemoteTabsTableViewController: UITableViewController {
     }
 
     fileprivate lazy var longPressRecognizer: UILongPressGestureRecognizer = {
-        return UILongPressGestureRecognizer(target: self, action: #selector(RemoteTabsTableViewController.longPress(_:)))
+        return UILongPressGestureRecognizer(target: self, action: #selector(RemoteTabsTableViewController.longPress))
```
Comment (jhugman) [inline Client/Frontend/Home/RemoteTabsPanel.swift]: Nit: remove `RemoteTabsTableViewController`?

File: Client/Frontend/Reader/ReaderModeStyleViewController.swift [hunk 10]
```diff
@@ -77,7 +77,7 @@ class ReaderModeStyleViewController: UIViewController {
             FontTypeButton(fontType: ReaderModeFontType.serif)
         ]
 
-        setupButtons(fontTypeButtons, inRow: fontTypeRow, action: #selector(ReaderModeStyleViewController.SELchangeFontType(_:)))
+        setupButtons(fontTypeButtons, inRow: fontTypeRow, action: #selector(ReaderModeStyleViewController.SELchangeFontType))
```
Comment (jhugman) [inline Client/Frontend/Reader/ReaderModeStyleViewController.swift]: Nit: remove `ReaderModeStyleViewController`?

File: Client/Frontend/Reader/ReaderModeStyleViewController.swift [hunk 11]
```diff
@@ -105,7 +105,7 @@ class ReaderModeStyleViewController: UIViewController {
             FontSizeButton(fontSizeAction: FontSizeAction.bigger)
         ]
 
-        setupButtons(fontSizeButtons, inRow: fontSizeRow, action: #selector(ReaderModeStyleViewController.SELchangeFontSize(_:)))
+        setupButtons(fontSizeButtons, inRow: fontSizeRow, action: #selector(ReaderModeStyleViewController.SELchangeFontSize))
```
Comment (jhugman) [inline Client/Frontend/Reader/ReaderModeStyleViewController.swift]: Nit: remove `ReaderModeStyleViewController`?

File: Client/Frontend/Reader/ReaderModeStyleViewController.swift [hunk 12]
```diff
@@ -124,7 +124,7 @@ class ReaderModeStyleViewController: UIViewController {
             ThemeButton(theme: ReaderModeTheme.sepia)
         ]
 
-        setupButtons(themeButtons, inRow: themeRow, action: #selector(ReaderModeStyleViewController.SELchangeTheme(_:)))
+        setupButtons(themeButtons, inRow: themeRow, action: #selector(ReaderModeStyleViewController.SELchangeTheme))
```
Comment (jhugman) [inline Client/Frontend/Reader/ReaderModeStyleViewController.swift]: Nit: remove `ReaderModeStyleViewController`?

File: Client/Frontend/Reader/ReaderModeStyleViewController.swift [hunk 13]
```diff
@@ -142,7 +142,7 @@ class ReaderModeStyleViewController: UIViewController {
         brightnessRow.addSubview(slider)
         slider.accessibilityLabel = NSLocalizedString("Brightness", comment: "Accessibility label for brightness adjustment slider in Reader Mode display settings")
         slider.tintColor = ReaderModeStyleViewControllerUX.BrightnessSliderTintColor
-        slider.addTarget(self, action: #selector(ReaderModeStyleViewController.SELchangeBrightness(_:)), for: UIControlEvents.valueChanged)
+        slider.addTarget(self, action: #selector(ReaderModeStyleViewController.SELchangeBrightness), for: .valueChanged)
```
Comment (jhugman) [inline Client/Frontend/Reader/ReaderModeStyleViewController.swift]: Nit: remove `ReaderModeStyleViewController`?

File: Client/Frontend/Settings/AppSettingsOptions.swift [hunk 14]
```diff
@@ -1012,7 +1012,7 @@ class StageSyncServiceDebugSetting: WithoutAccountSetting {
         super.onConfigureCell(cell)
         let control = UISwitch()
         control.onTintColor = UIConstants.ControlTintColor
-        control.addTarget(self, action: #selector(StageSyncServiceDebugSetting.switchValueChanged(_:)), for: UIControlEvents.valueChanged)
+        control.addTarget(self, action: #selector(StageSyncServiceDebugSetting.switchValueChanged), for: .valueChanged)
```
Comment (jhugman) [inline Client/Frontend/Settings/AppSettingsOptions.swift]: Nit: remove `StageSyncServiceDebugSetting`?

File: Client/Frontend/Settings/ClearPrivateDataTableViewController.swift [hunk 15]
```diff
@@ -61,21 +61,21 @@ class ClearPrivateDataTableViewController: UITableViewController {
...
     override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        let cell = UITableViewCell(style: UITableViewCellStyle.default, reuseIdentifier: nil)
+        let cell = UITableViewCell(style: .default, reuseIdentifier: nil)
 
         if indexPath.section == SectionToggles {
             cell.textLabel?.text = clearables[indexPath.item].clearable.label
             let control = UISwitch()
             control.onTintColor = UIConstants.ControlTintColor
-            control.addTarget(self, action: #selector(ClearPrivateDataTableViewController.switchValueChanged(_:)), for: UIControlEvents.valueChanged)
+            control.addTarget(self, action: #selector(ClearPrivateDataTableViewController.switchValueChanged), for: .valueChanged)
```
Comment (jhugman) [inline Client/Frontend/Settings/ClearPrivateDataTableViewController.swift]: Nit: remove `ClearPrivateDataTableViewController`?

File: Client/Frontend/Settings/SearchSettingsTableViewController.swift [hunk 16]
```diff
@@ -84,20 +84,20 @@ class SearchSettingsTableViewController: UITableViewController {
...
                 cell.imageView?.layer.cornerRadius = 4
                 cell.imageView?.layer.masksToBounds = true
             case ItemDefaultSuggestions:
-                cell = UITableViewCell(style: UITableViewCellStyle.default, reuseIdentifier: nil)
+                cell = UITableViewCell(style: .default, reuseIdentifier: nil)
                 cell.textLabel?.text = NSLocalizedString("Show Search Suggestions", comment: "Label for show search suggestions setting.")
                 let toggle = UISwitch()
                 toggle.onTintColor = UIConstants.ControlTintColor
-                toggle.addTarget(self, action: #selector(SearchSettingsTableViewController.didToggleSearchSuggestions(_:)), for: UIControlEvents.valueChanged)
+                toggle.addTarget(self, action: #selector(SearchSettingsTableViewController.didToggleSearchSuggestions), for: .valueChanged)
```
Comment (jhugman) [inline Client/Frontend/Settings/SearchSettingsTableViewController.swift]: Nit: remove `SearchSettingsTableViewController`?

File: Client/Frontend/Widgets/SearchInputView.swift [hunk 17]
```diff
@@ -40,7 +40,7 @@ class SearchInputView: UIView {
         textField.delegate = self
         textField.textColor = SearchInputViewUX.inputColor
         textField.tintColor = SearchInputViewUX.inputColor
-        textField.addTarget(self, action: #selector(SearchInputView.inputTextDidChange(_:)), for: .editingChanged)
+        textField.addTarget(self, action: #selector(SearchInputView.inputTextDidChange), for: .editingChanged)
```
Comment (jhugman) [inline Client/Frontend/Widgets/SearchInputView.swift]: Nit: remove `SearchInputView`?

File: ReadingList/ReadingListRecordResponse.swift [hunk 18]
```diff
@@ -17,12 +17,6 @@ class ReadingListRecordResponse: ReadingListResponse {
...
 
     var record: ReadingListServerRecord? {
-        get {
-            if let json = self.json {
-                return ReadingListServerRecord(json: json)
-            } else {
-                return nil
-            }
-        }
+        return json.flatMap { ReadingListServerRecord(json: $0) }
```
Comment (jhugman) [inline ReadingList/ReadingListRecordResponse.swift]: This sort of re-write, while welcome, should probably be in its own PR.

File: Shared/DeviceInfo.swift [hunk 19]
```diff
@@ -61,9 +61,7 @@ open class DeviceInfo {
     open class func hasConnectivity() -> Bool {
         let status = Reach().connectionStatus()
         switch status {
-        case .online(.wwan):
-            return true
-        case .online(.wiFi):
+        case .online(.wwan), .online(.wiFi):
```
Comment (jhugman) [inline Shared/DeviceInfo.swift]: These code cleanups should probably be in their own PR.

File: Shared/Extensions/NSFileManagerExtensions.swift [hunk 20]
```diff
@@ -60,10 +56,10 @@ public extension FileManager {
     }
 
     func allocatedSizeOfDirectoryAtURL(_ url: URL, forFilesPrefixedWith prefix: String, isLargerThanBytes threshold: Int64) throws -> Bool {
-        let directoryEnumerator = try directoryEnumeratorForURL(url)
+        let enumerator = try directoryEnumerator(for: url)
         var acc: Int64 = 0
-        for item in directoryEnumerator {
-            acc += try sizeForItemURL(item as AnyObject, withPrefix: prefix)
+        for item in enumerator {
+            acc += try size(for: item as AnyObject, withPrefix: prefix)
```
Comment (jhugman) [inline Shared/Extensions/NSFileManagerExtensions.swift]: These code cleanups should probably be in their own PR.

File: Shared/Extensions/NSFileManagerExtensions.swift [hunk 21]
```diff
@@ -80,9 +76,9 @@ public extension FileManager {
      - throws: Error reading/operating on disk.
      */
     func getAllocatedSizeOfDirectoryAtURL(_ url: URL, forFilesPrefixedWith prefix: String) throws -> Int64 {
-        let directoryEnumerator = try directoryEnumeratorForURL(url)
-        return try directoryEnumerator.reduce(0) {
-            let size = try sizeForItemURL($1 as AnyObject, withPrefix: prefix)
+        let enumerator = try directoryEnumerator(for: url)
+        return try enumerator.reduce(0) {
+            let size = try self.size(for: $1 as AnyObject, withPrefix: prefix)
```
Comment (jhugman) [inline Shared/Extensions/NSFileManagerExtensions.swift]: These code cleanups should probably be in their own PR.

File: Shared/Extensions/NSStringExtensions.swift [hunk 22]
```diff
@@ -6,14 +6,8 @@ import Foundation
...
-        if let path = bundle.path(forResource: name, ofType: type) {
-            do {
-                return try String(contentsOfFile: path, encoding: encoding)
-            } catch {
-                return nil
-            }
-        } else {
-            return nil
+        return bundle.path(forResource: name, ofType: type).flatMap {
+            try? String(contentsOfFile: $0, encoding: encoding)
```
Comment (jhugman) [inline Shared/Extensions/NSStringExtensions.swift]: These code cleanups should probably be in their own PR.

File: Shared/Extensions/NSURLExtensions.swift [hunk 23]
```diff
@@ -162,10 +162,7 @@ extension URL {
     }
 
     public var normalizedHostAndPath: String? {
-        if let normalizedHost = self.normalizedHost {
-            return normalizedHost + self.path
-        }
-        return nil
+        return normalizedHost.flatMap { $0 + self.path }
```
Comment (jhugman) [inline Shared/Extensions/NSURLExtensions.swift]: These code cleanups should probably be in their own PR.

File: Shared/Extensions/NSURLExtensions.swift [hunk 24]
```diff
@@ -267,20 +260,12 @@ extension URL {
     :returns: The public suffix for within the given hostname.
     */
     public var publicSuffix: String? {
-        if let host = self.host {
-            return publicSuffixFromHost(host, withAdditionalParts: 0)
-        } else {
-            return nil
-        }
+        return host.flatMap { publicSuffixFromHost($0, withAdditionalParts: 0) }
```
Comment (jhugman) [inline Shared/Extensions/NSURLExtensions.swift]: These code cleanups should probably be in their own PR.

File: Shared/Extensions/NSURLExtensions.swift [hunk 25]
```diff
@@ -194,11 +191,7 @@ extension URL {
         }
 
         if self.isErrorPageURL {
-            if let decodedURL = self.originalURLFromErrorURL {
-                return decodedURL.displayURL
-            } else {
-                return nil
-            }
+            return originalURLFromErrorURL?.displayURL
```
Comment (jhugman) [inline Shared/Extensions/NSURLExtensions.swift]: These code cleanups should probably be in their own PR.

File: Shared/Extensions/NSURLExtensions.swift [hunk 26]
```diff
@@ -267,20 +260,12 @@ extension URL {
...
     }
 
     public func isWebPage(includeDataURIs: Bool = true) -> Bool {
         let schemes = includeDataURIs ? ["http", "https", "data"] : ["http", "https"]
-        if let scheme = scheme, schemes.contains(scheme) {
-            return true
-        }
-
-        return false
+        return scheme.map { schemes.contains($0) } ?? false
```
Comment (jhugman) [inline Shared/Extensions/NSURLExtensions.swift]: These code cleanups should probably be in their own PR.

File: Client/Frontend/Browser/ButtonToast.swift (hunk 4 above)
Comment (adamnemecek) [reply Client/Frontend/Browser/ButtonToast.swift]: Those where I just remove (_:) were done with a search and replace. I will remove the class names in the next PR. This one was getting a little bloated.

File: ReadingList/ReadingListRecordResponse.swift (hunk 18 above)
Comment (adamnemecek) [reply ReadingList/ReadingListRecordResponse.swift]: It probably should but if you could make an exception this one time, I would be really grateful. I've had this code for the last months and I would just kind of want to merge it in so that I can proceed with the rest.
